x/trigger/keeper/event_detector_test.go

```diff
@@ -215,10 +261,14 @@ func (s *KeeperTestSuite) TestDetectBlockEvents() {
 				s.ctx.GasMeter().RefundGas(s.ctx.GasMeter().GasConsumed(), "testing")
 				registered = append(registered, trigger)
 			}
+			s.ctx = s.ctx.WithEventManager(sdk.NewEventManagerWithHistory(s.ctx.EventManager().GetABCIEventHistory()))
 
 			// Action
 			s.app.TriggerKeeper.DetectBlockEvents(s.ctx)
 
+			events := s.ctx.EventManager().Events()
+			s.Equal(tc.events, events, "should have correct events from detected events in DetectBlockEvents")
```

Comment:
No need to change. But if there's other changes to make to the PR, it might be a good change to make. Feel free to click "resolved" on this comment.

Since the event attributes are all `[]byte`, they can be hard to recognize in the output of a failed `.Equal` check. In the `"github.com/provenance-io/provenance/testutil/assertions"` package, there's `AssertEqualEvents` which really helps with this.

```suggestion
			assertions.AssertEqualEvents(s.T(), tc.events, events, "should have correct events from detected events in DetectBlockEvents")
```

---

With `.Equal` it can be hard to identify exactly which one of the entries changed. And if a new event shows up, it's difficult to tell where in the list it was added.

I threw together a sample where one attribute value changes, and a new event is added.

Here's the expected:
```golang
	expEvents := sdk.Events{
		{
			Type: "eventFoo",
			Attributes: []abci.EventAttribute{
				{Key: []byte("someattr"), Value: []byte(addr1)},
				{Key: []byte("athing"), Value: []byte("88"), Index: true},
			},
		},
		{
			Type: "eventBar",
			Attributes: []abci.EventAttribute{
				{Key: []byte("fielda"), Value: []byte(addr2)},
				{Key: []byte("anotherkey"), Value: []byte("anotherval")},
			},
		},
		{
			Type: "eventBaz",
			Attributes: []abci.EventAttribute{
				{Key: []byte("something"), Value: []byte("abcdefg")},
			},
		},
	}
```

The failure message using `.Equal`:
```plaintext
        	Error Trace:	/Users/danielwedul/git/provenance/x/exchange/keeper/keys_test.go:3858
        	Error:      	Not equal: 
        	            	expected: types.Events{types.Event{Type:"eventFoo", Attributes:[]types.EventAttribute{types.EventAttribute{Key:[]uint8{0x73, 0x6f, 0x6d, 0x65, 0x61, 0x74, 0x74, 0x72}, Value:[]uint8{0x70, 0x62, 0x31, 0x67, 0x34, 0x75, 0x78, 0x7a, 0x6d, 0x6e, 0x73, 0x64, 0x33, 0x6a, 0x35, 0x7a, 0x65, 0x72, 0x79, 0x77, 0x66, 0x30, 0x34, 0x37, 0x68, 0x36, 0x6c, 0x74, 0x61, 0x30, 0x34, 0x37, 0x68, 0x36, 0x6c, 0x39, 0x65, 0x77, 0x67, 0x7a, 0x38}, Index:false}, types.EventAttribute{Key:[]uint8{0x61, 0x74, 0x68, 0x69, 0x6e, 0x67}, Value:[]uint8{0x38, 0x38}, Index:true}}}, types.Event{Type:"eventBar", Attributes:[]types.EventAttribute{types.EventAttribute{Key:[]uint8{0x66, 0x69, 0x65, 0x6c, 0x64, 0x61}, Value:[]uint8{0x70, 0x62, 0x31, 0x67, 0x34, 0x75, 0x78, 0x7a, 0x6d, 0x74, 0x73, 0x64, 0x33, 0x6a, 0x35, 0x7a, 0x65, 0x72, 0x79, 0x77, 0x66, 0x30, 0x34, 0x37, 0x68, 0x36, 0x6c, 0x74, 0x61, 0x30, 0x34, 0x37, 0x68, 0x36, 0x6c, 0x79, 0x63, 0x6d, 0x7a, 0x77, 0x65}, Index:false}, types.EventAttribute{Key:[]uint8{0x61, 0x6e, 0x6f, 0x74, 0x68, 0x65, 0x72, 0x6b, 0x65, 0x79}, Value:[]uint8{0x61, 0x6e, 0x6f, 0x74, 0x68, 0x65, 0x72, 0x76, 0x61, 0x6c}, Index:false}}}, types.Event{Type:"eventBaz", Attributes:[]types.EventAttribute{types.EventAttribute{Key:[]uint8{0x73, 0x6f, 0x6d, 0x65, 0x74, 0x68, 0x69, 0x6e, 0x67}, Value:[]uint8{0x61, 0x62, 0x63, 0x64, 0x65, 0x66, 0x67}, Index:false}}}}
        	            	actual  : types.Events{types.Event{Type:"eventFoo", Attributes:[]types.EventAttribute{types.EventAttribute{Key:[]uint8{0x73, 0x6f, 0x6d, 0x65, 0x61, 0x74, 0x74, 0x72}, Value:[]uint8{0x70, 0x62, 0x31, 0x67, 0x34, 0x75, 0x78, 0x7a, 0x6d, 0x6e, 0x73, 0x64, 0x33, 0x6a, 0x35, 0x7a, 0x65, 0x72, 0x79, 0x77, 0x66, 0x30, 0x34, 0x37, 0x68, 0x36, 0x6c, 0x74, 0x61, 0x30, 0x34, 0x37, 0x68, 0x36, 0x6c, 0x39, 0x65, 0x77, 0x67, 0x7a, 0x38}, Index:false}, types.EventAttribute{Key:[]uint8{0x61, 0x74, 0x68, 0x69, 0x6e, 0x67}, Value:[]uint8{0x38, 0x38}, Index:true}}}, types.Event{Type:"eventBar", Attributes:[]types.EventAttribute{types.EventAttribute{Key:[]uint8{0x66, 0x69, 0x65, 0x6c, 0x64, 0x61}, Value:[]uint8{0x74, 0x70, 0x31, 0x67, 0x34, 0x75, 0x78, 0x7a, 0x6d, 0x74, 0x73, 0x64, 0x33, 0x6a, 0x35, 0x7a, 0x65, 0x72, 0x79, 0x77, 0x66, 0x30, 0x34, 0x37, 0x68, 0x36, 0x6c, 0x74, 0x61, 0x30, 0x34, 0x37, 0x68, 0x36, 0x6c, 0x68, 0x6e, 0x37, 0x77, 0x76, 0x6e}, Index:false}, types.EventAttribute{Key:[]uint8{0x61, 0x6e, 0x6f, 0x74, 0x68, 0x65, 0x72, 0x6b, 0x65, 0x79}, Value:[]uint8{0x61, 0x6e, 0x6f, 0x74, 0x68, 0x65, 0x72, 0x76, 0x61, 0x6c}, Index:false}}}, types.Event{Type:"eventNew", Attributes:[]types.EventAttribute{types.EventAttribute{Key:[]uint8{0x74, 0x79, 0x70, 0x65, 0x33, 0x2d, 0x61, 0x74, 0x74, 0x72, 0x31}, Value:[]uint8{0x74, 0x79, 0x70, 0x65, 0x33, 0x2d, 0x76, 0x61, 0x6c, 0x31}, Index:false}, types.EventAttribute{Key:[]uint8{0x74, 0x79, 0x70, 0x65, 0x33, 0x2d, 0x61, 0x74, 0x74, 0x72, 0x32}, Value:[]uint8{0x74, 0x79, 0x70, 0x65, 0x33, 0x2d, 0x76, 0x61, 0x6c, 0x32}, Index:false}}}, types.Event{Type:"eventBaz", Attributes:[]types.EventAttribute{types.EventAttribute{Key:[]uint8{0x73, 0x6f, 0x6d, 0x65, 0x74, 0x68, 0x69, 0x6e, 0x67}, Value:[]uint8{0x61, 0x62, 0x63, 0x64, 0x65, 0x66, 0x67}, Index:false}}}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-(types.Events) (len=3) {
        	            	+(types.Events) (len=4) {
        	            	  (types.Event) {
        	            	@@ -34,5 +34,5 @@
        	            	     Value: ([]uint8) (len=41) {
        	            	-     00000000  70 62 31 67 34 75 78 7a  6d 74 73 64 33 6a 35 7a  |pb1g4uxzmtsd3j5z|
        	            	+     00000000  74 70 31 67 34 75 78 7a  6d 74 73 64 33 6a 35 7a  |tp1g4uxzmtsd3j5z|
        	            	      00000010  65 72 79 77 66 30 34 37  68 36 6c 74 61 30 34 37  |erywf047h6lta047|
        	            	-     00000020  68 36 6c 79 63 6d 7a 77  65                       |h6lycmzwe|
        	            	+     00000020  68 36 6c 68 6e 37 77 76  6e                       |h6lhn7wvn|
        	            	     },
        	            	@@ -46,2 +46,25 @@
        	            	      00000000  61 6e 6f 74 68 65 72 76  61 6c                    |anotherval|
        	            	+    },
        	            	+    Index: (bool) false
        	            	+   }
        	            	+  }
        	            	+ },
        	            	+ (types.Event) {
        	            	+  Type: (string) (len=8) "eventNew",
        	            	+  Attributes: ([]types.EventAttribute) (len=2) {
        	            	+   (types.EventAttribute) {
        	            	+    Key: ([]uint8) (len=11) {
        	            	+     00000000  74 79 70 65 33 2d 61 74  74 72 31                 |type3-attr1|
        	            	+    },
        	            	+    Value: ([]uint8) (len=10) {
        	            	+     00000000  74 79 70 65 33 2d 76 61  6c 31                    |type3-val1|
        	            	+    },
        	            	+    Index: (bool) false
        	            	+   },
        	            	+   (types.EventAttribute) {
        	            	+    Key: ([]uint8) (len=11) {
        	            	+     00000000  74 79 70 65 33 2d 61 74  74 72 32                 |type3-attr2|
        	            	+    },
        	            	+    Value: ([]uint8) (len=10) {
        	            	+     00000000  74 79 70 65 33 2d 76 61  6c 32                    |type3-val2|
        	            	     },
        	Test:       	TestFoo/equal
```

The using `AssertEqualEvents`:
```plaintext
        	Error Trace:	/Users/danielwedul/git/provenance/testutil/assertions/events.go:61
        	            				/Users/danielwedul/git/provenance/x/exchange/keeper/keys_test.go:3861
        	Error:      	Not equal: 
        	            	expected: []string{"[0]eventFoo[0]: \"someattr\" = \"pb1g4uxzmnsd3j5zerywf047h6lta047h6l9ewgz8\"", "[0]eventFoo[1]: \"athing\" = \"88\" (indexed)", "[1]eventBar[0]: \"fielda\" = \"pb1g4uxzmtsd3j5zerywf047h6lta047h6lycmzwe\"", "[1]eventBar[1]: \"anotherkey\" = \"anotherval\"", "[2]eventBaz[0]: \"something\" = \"abcdefg\""}
        	            	actual  : []string{"[0]eventFoo[0]: \"someattr\" = \"pb1g4uxzmnsd3j5zerywf047h6lta047h6l9ewgz8\"", "[0]eventFoo[1]: \"athing\" = \"88\" (indexed)", "[1]eventBar[0]: \"fielda\" = \"tp1g4uxzmtsd3j5zerywf047h6lta047h6lhn7wvn\"", "[1]eventBar[1]: \"anotherkey\" = \"anotherval\"", "[2]eventNew[0]: \"type3-attr1\" = \"type3-val1\"", "[2]eventNew[1]: \"type3-attr2\" = \"type3-val2\"", "[3]eventBaz[0]: \"something\" = \"abcdefg\""}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,7 +1,9 @@
        	            	-([]string) (len=5) {
        	            	+([]string) (len=7) {
        	            	  (string) (len=72) "[0]eventFoo[0]: \"someattr\" = \"pb1g4uxzmnsd3j5zerywf047h6lta047h6l9ewgz8\"",
        	            	  (string) (len=41) "[0]eventFoo[1]: \"athing\" = \"88\" (indexed)",
        	            	- (string) (len=70) "[1]eventBar[0]: \"fielda\" = \"pb1g4uxzmtsd3j5zerywf047h6lta047h6lycmzwe\"",
        	            	+ (string) (len=70) "[1]eventBar[0]: \"fielda\" = \"tp1g4uxzmtsd3j5zerywf047h6lta047h6lhn7wvn\"",
        	            	  (string) (len=43) "[1]eventBar[1]: \"anotherkey\" = \"anotherval\"",
        	            	- (string) (len=39) "[2]eventBaz[0]: \"something\" = \"abcdefg\""
        	            	+ (string) (len=44) "[2]eventNew[0]: \"type3-attr1\" = \"type3-val1\"",
        	            	+ (string) (len=44) "[2]eventNew[1]: \"type3-attr2\" = \"type3-val2\"",
        	            	+ (string) (len=39) "[3]eventBaz[0]: \"something\" = \"abcdefg\""
        	            	 }
        	Test:       	TestFoo/assertions
```